redis/src/cluster_async/mod.rs

```diff
@@ -446,56 +450,89 @@ where
-            let mut write_guard = inner.conn_lock.write().await;
-            let mut connections = mem::take(&mut write_guard.0);
-            let slots = &mut write_guard.1;
-            let mut result = Ok(());
-            for (_, conn) in connections.iter_mut() {
-                let mut conn = conn.clone().await;
-                let value = match conn.req_packed_command(&slot_cmd()).await {
-                    Ok(value) => value,
-                    Err(err) => {
-                        result = Err(err);
-                        continue;
-                    }
-                };
-                match parse_slots(value, inner.cluster_params.tls).and_then(|v| {
-                    Self::build_slot_map(slots, v, inner.cluster_params.read_from_replicas)
-                }) {
+            let read_guard = inner.conn_lock.write().await;
+            let num_of_nodes = read_guard.0.len();
+            let amount = std::cmp::min(num_of_nodes, MAX_REQUESTED_NODES);
+            let mut requested_nodes = {
+                let mut rng = thread_rng();
+                read_guard.0.values().choose_multiple(&mut rng, amount)
+            };
+            let mut cluster_slot_futures = Vec::new();
+            for conn in requested_nodes.iter_mut() {
+                let mut conn: C = conn.clone().await;
+                let slots_future = async move { conn.req_packed_command(&slot_cmd()).await };
+                #[cfg(feature = "tokio-comp")]
+                cluster_slot_futures.push(tokio::spawn(slots_future));
+                #[cfg(all(not(feature = "tokio-comp"), feature = "async-std-comp"))]
+                cluster_slot_futures.push(AsyncStd::spawn(slots_future));
+            }
+            let mut topology_join_results = futures::future::join_all(cluster_slot_futures).await;
+            let mut topology_results: Vec<_> = topology_join_results
+                .drain(..)
+                .filter_map(|r| r.ok())
+                .collect();
+            topology_results.shuffle(&mut thread_rng());
+            let topology_vec = calculate_topology(topology_results, retries.clone())?;
+            let mut slots = SlotMap::new();
+            for (idx, topology_view) in topology_vec.iter().enumerate() {
+                match parse_slots(&topology_view.topology_value, inner.cluster_params.tls).and_then(
+                    |v| {
+                        Self::build_slot_map(&mut slots, v, inner.cluster_params.read_from_replicas)
+                    },
+                ) {
                     Ok(_) => {
-                        result = Ok(());
                         break;
                     }
-                    Err(err) => result = Err(err),
+                    Err(e) => {
+                        // If it's the last view, raise the error.
+                        if idx == topology_vec.len() - 1 {
+                            return Err(e);
+                        } else {
+                            continue;
```

Comment:
so, while we want to use the best view, we're fine with using a view that is held by the minority of nodes?

Reply:
If it's the last retry AND we still couldn't find a majority, yes, we will be ok by taking a view with full coverage